src/commands/publish/preview/mod.rs

```diff
@@ -10,63 +10,158 @@ use uuid::Uuid;
+    let client: Client;
+
+    let preview = match &user {
+        Some(user) => {
+            log::info!("Running in Auth'd mode");
+
+            project.validate(false)?;
+
+            commands::build(&project)?;
+            client = http::auth_client(&user);
+
+            authenticated_upload(&client, &project)?
+        }
+        None => {
+            log::info!("Running in Un-auth'd mode");
+
+            // KV namespaces are not supported by the preview service unless you authenticate
+            // so we omit them and provide the user with a little guidance. We don't error out, though,
+            // because there are valid workarounds for this for testing purposes.
+            if project.kv_namespaces.is_some() {
+                message::warn("KV Namespaces are not supported in unauthenticated mode");
+                message::help(
+                    "Run wrangler config or set $CF_API_KEY and $CF_EMAIL to configure your user.",
+                );
+                project.kv_namespaces = None;
+            }
+
+            commands::build(&project)?;
+            client = http::client();
+
+            unauthenticated_upload(&client, &project)?
+        }
+    };
+
+    let worker_res = call_worker(&client, preview, method, body)?;
+
+    let msg = format!("Your worker responded with: {}", worker_res);
+    message::preview(&msg);
 
-    let client = http::client();
+    Ok(())
+}
 
-    commands::build(&project)?;
+fn authenticated_upload(client: &Client, project: &Project) -> Result<Preview, failure::Error> {
+    let create_address = format!(
+        "https://api.cloudflare.com/client/v4/accounts/{}/workers/scripts/{}/preview",
+        project.account_id, project.name
+    );
+    log::info!("address: {}", create_address);
+
+    let script_upload_form = publish::build_script_upload_form(&project)?;
+
+    let mut res = client
+        .post(&create_address)
+        .multipart(script_upload_form)
+        .send()?
+        .error_for_status()?;
+
+    let text = &res.text()?;
+    log::info!("Response from preview: {:#?}", text);
+
+    let response: V4ApiResponse =
+        serde_json::from_str(text).expect("could not create a script on cloudflareworkers.com");
+
+    Ok(Preview::from(response.result))
+}
+
+fn unauthenticated_upload(client: &Client, project: &Project) -> Result<Preview, failure::Error> {
+    let create_address = "https://cloudflareworkers.com/script";
+    log::info!("address: {}", create_address);
 
     let script_upload_form = publish::build_script_upload_form(project)?;
 
-    let res = client
+    let mut res = client
         .post(create_address)
         .multipart(script_upload_form)
         .send()?
-        .error_for_status();
+        .error_for_status()?;
 
-    let text = &res?.text()?;
-    log::info!("Response from preview: {:?}", text);
+    let text = &res.text()?;
+    log::info!("Response from preview: {:#?}", text);
```

Comment:
🔥

Reply:
yeah this only mattered bc the response is embedded in a v4 struct when it goes through the API